Include/Rover/Box.hpp

```diff
@@ -10,13 +12,91 @@ namespace Rover {
     \tparam T The type of value to generate.
   */
   template<typename T>
-  class Box : private Noncopyable {
+  class Box {
     public:
 
       //! The type of value to generate.
       using Type = T;
 
-      Type generate(Evaluator& evaluator) const;
+      //! Packs a generator or generator pointer into a Box
+      template<typename GeneratorFwd, std::enable_if_t<!std::is_same_v<
+          std::decay_t<GeneratorFwd>, Box>>* = nullptr>
+      explicit Box(GeneratorFwd&& gen);
+
+      Type generate(Evaluator& evaluator);
+
+    private:
+
+      struct WrapperBase {
+        virtual ~WrapperBase() = default;
+        virtual Type generate(Evaluator& evaluator) = 0;
+      };
+
+      template<typename Generator>
+      struct ValueWrapper;
+
+      template<typename GeneratorFwd>
+      ValueWrapper(GeneratorFwd&&) -> ValueWrapper<std::decay_t<GeneratorFwd>>;
+
+      template<typename Generator>
+      struct PointerWrapper;
+
+      template<typename GeneratorFwd>
+      PointerWrapper(GeneratorFwd&&)->PointerWrapper<std::decay_t<GeneratorFwd>>;
+
+      std::shared_ptr<WrapperBase> m_generator;
```

Comment:
Haven't found a better way as we should be able to copy Boxes. Using unique or raw pointers, I could not recover the original type of the copied-from Box.